### PR TITLE
Automatic detection of target versions for local upgrade

### PR DESCRIPTION
**Pull request by M0rgan01**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Automatic detection of target versions for local upgrade
| Type?             | improvement
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | -
| Sponsor company   | -
| How to test?      | A local upgrade should work the same as before

### CODE REVIEW

File: classes/Exceptions/UpgradeException.php [hunk 1]
```diff
@@ -25,7 +25,7 @@
  * @license   https://opensource.org/licenses/AFL-3.0 Academic Free License 3.0 (AFL-3.0)
  */
 
-namespace PrestaShop\Module\AutoUpgrade;
```
Comment (Quetzacoalt91) [inline classes/Exceptions/UpgradeException.php:28]: On #715, the new namespace is `Exception` in the singular form

Comment (Quetzacoalt91) [reply classes/Exceptions/UpgradeException.php:28]: Fixed with https://github.com/PrestaShop/autoupgrade/pull/715/commits/813724a52a9b942bc017e1f8e550f32b83a3982f

File: classes/TaskRunner/Miscellaneous/UpdateConfig.php [hunk 2]
```diff
@@ -174,4 +182,57 @@ private function writeConfig(array $config): bool
...
+     * @throws ZipActionException
+     * @throws Exception
+     */
+    private function extractPrestashopVersionFromZip(string $zipFile): string
+    {
+        $internalZipFileName = 'prestashop.zip';
+        $versionFile = 'install/install_version.php';
+
+        if (!file_exists($zipFile)) {
+            throw new FileNotFoundException("Unable to find $zipFile file");
```
Comment (Quetzacoalt91) [inline classes/TaskRunner/Miscellaneous/UpdateConfig.php:196]: In think these exceptions messages might be not visible in the logs. Is that expected?

File: classes/TaskRunner/Miscellaneous/UpdateConfig.php [hunk 3]
```diff
@@ -174,4 +182,57 @@ private function writeConfig(array $config): bool
...
+        $internalZipContent = $this->container->getZipAction()->extractFileFromArchive($zip, $internalZipFileName);
+        $zip->close();
+
+        $tempInternalZipPath = $this->createTemporaryFile($internalZipContent);
+
+        $internalZip = $this->container->getZipAction()->open($tempInternalZipPath);
+        $fileContent = $this->container->getZipAction()->extractFileFromArchive($internalZip, $versionFile);
+        $internalZip->close();
+
+        unlink($tempInternalZipPath);
```
Comment (Quetzacoalt91) [inline classes/TaskRunner/Miscellaneous/UpdateConfig.php:208]: > Deletes filename. Similar to the Unix C unlink() function. An [E_WARNING](https://www.php.net/manual/en/errorfunc.constants.php#constant.e-warning) level error will be generated on failure.

File: classes/ZipAction.php [hunk 4]
```diff
@@ -238,20 +240,32 @@ private function isFileWithinFileSizeLimit(string $filepath): bool
...
-        $zip = new \ZipArchive();
+        $zip = new ZipArchive();
         if (null === $flags) {
             $flags = 0;
         }
         if ($zip->open($zipFile, $flags) !== true || empty($zip->filename)) {
             $this->logger->error($this->translator->trans('Unable to open zipFile %s', [$zipFile]));
-
-            return false;
+            throw new ZipActionException("Unable to open $zipFile file");
```
Comment (Quetzacoalt91) [inline classes/ZipAction.php:253]: Can we use the same content on both lines?

File: classes/ZipAction.php [hunk 5]
```diff
@@ -238,20 +240,32 @@ private function isFileWithinFileSizeLimit(string $filepath): bool
...
+     * @throws ZipActionException
+     */
+    public function extractFileFromArchive(ZipArchive $zip, string $fileName): string
+    {
+        $fileIndex = $zip->locateName($fileName);
+        if ($fileIndex !== false) {
+            return $zip->getFromIndex($fileIndex);
+        } else {
+            throw new ZipActionException("Unable to find $fileName file");
+        }
```
Comment (Quetzacoalt91) [inline classes/ZipAction.php:269]:
```suggestion
        }

        throw new ZipActionException("Unable to find $fileName file");

```

File: classes/TaskRunner/Miscellaneous/UpdateConfig.php [hunk 6]
```diff
@@ -174,4 +182,57 @@ private function writeConfig(array $config): bool
...
+    /**
+     * @throws ZipActionException
+     */
+    private function extractVersionFromContent(string $content): string
+    {
+        $pattern = "/define\\('_PS_INSTALL_VERSION_', '([\\d.]+)'\\);/";
+        if (preg_match($pattern, $content, $matches)) {
+            return $matches[1];
+        } else {
+            throw new ZipActionException('Unable to extract version from content');
```
Comment (Quetzacoalt91) [inline classes/TaskRunner/Miscellaneous/UpdateConfig.php:235]: Is the type `ZipActionException` appropriate here?

File: classes/TaskRunner/Miscellaneous/UpdateConfig.php (hunk 2 above)
Comment (M0rgan01) [reply classes/TaskRunner/Miscellaneous/UpdateConfig.php:196]: Well seen, you have to change the error log to replace it with the exception potentially raised

File: classes/TaskRunner/Miscellaneous/UpdateConfig.php (hunk 6 above)
Comment (M0rgan01) [reply classes/TaskRunner/Miscellaneous/UpdateConfig.php]: Maybe not

**Review by Quetzacoalt91 — APPROVED** [review]

**Comment by AureRita** [comment]
Hi @M0rgan01 

Thank you for your PR, I tested it and : 

on 1.7.1.0, when we click on save button to load the version, the module doesn't do anything as you can see : 

[recording(235).webm](https://github.com/PrestaShop/autoupgrade/assets/100699445/7abd7a6a-25b3-4e9d-b9ad-7eff6fa87fd4)


On 1.7.7.8 and other version, you can load the version but in the information of the version, you'll keep the version N/A as you can see : 

[recording(236).webm](https://github.com/PrestaShop/autoupgrade/assets/100699445/4c11133b-22f7-4331-ba76-2f39baefcbe5)

After that, I do some non regrression test from : 

8.0.5 to 9.0.0 :heavy_check_mark: 
8.1.6 to 9.0.0 :heavy_check_mark: 
8.1.5 to 9.0.0 :heavy_check_mark: 
8.0.5 to 8.1.6 :heavy_check_mark: 
8.1.5 to 8.1.6 :heavy_check_mark: 

Waiting for your correction

**Comment by M0rgan01** [comment]
Hello @AureRita , on a 1.7.1.0, the functionality works for me, please empty the browser cache and check that the module is correctly up to date during the test
[Capture vidéo du 18-06-2024 16:41:51.webm](https://github.com/PrestaShop/autoupgrade/assets/36233446/1a94f85f-767a-408f-8a95-0d7f8f4b5968)

Concerning the "Latest official version for archive channel.: N/D" part, it is not a regression, it is already like that on the dev branch. If this is not intended behavior, an issue should be created for it

[Capture vidéo du 18-06-2024 16:55:12.webm](https://github.com/PrestaShop/autoupgrade/assets/36233446/a18c5089-4494-4ead-8d3c-c5cd33e0dbf6)

**Pull request merged**